### PR TITLE
17419 Date range picker in place editor

### PR DESCRIPTION
**Pull request by knight-bubble**
[`* `#18939` Date picker inplace editor`](http://community.openproject.org/work_packages/18939)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://hirtie-maxim-feature-17419-date-picker-in-place-editor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://hirtie-maxim-feature-17419-date-picker-in-place-editor.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by ulferts** [comment]
@hirtie-maxim I tried the element and I like the behaviour but there seems to be one limitation/bug in that I cannot switch month in order to set a range that spans more than one month.

When I switch the month, the start date is switched to be in that month as well. If it is not possible to work around that limitation, the UI is not suitable.

File: frontend/public/templates/components/inplace_editor/display/daterange.html [hunk 1]
```diff
@@ -0,0 +1 @@
+<op-date date-value="dataObject.startDate"></op-date> - <op-date date-value="dataObject.dueDate"></op-date>
```
Comment (0xF013) [inline frontend/public/templates/components/inplace_editor/display/daterange.html]: @hirtie-maxim linebreak please

**Comment by ulferts** [comment]
@hirtie-maxim not looking at the code in my review I noticed the following:
- cannot change first day of the week (Sunday, Saturday, Monday). This should be configurable via the setting in "Modules" -> "Administration" -> "Settings" -> "Display"
- cannot set the values to null. Once you have start/end date set to anything, you can not choose to set them to null again.
- When no date is set, the value is displayed as "Invalid date - invalid date". This should read "No start date set - No end date set". 
- I can always only set ranges, not individual dates for example when I only want to specify the start or end date. This might break the idea of having one field with a date range picker.
- The calendar week is not displayed.
- I checked with PM and the requirement is to always display the dates in the format YYYY/MM/DD in while in edit mode. When in show mode, the display should be according to the settings in the admin section. Currently this admin setting is used for both which I personally find more desirable.

**Comment by floriank** [comment]
@hirtie-maxim there is still a breaking test in the `spec` group, which seems to be flickering (or caused by a solar flare)

 I restarted the build, let's wait and see, as I doubt your changes broke it. I'd like you to have a look at the @houndci comments though.

**Comment by 0xF013** [comment]
@floriank we can always xit the test

**Comment by floriank** [comment]
@0xF013 Yeah maybe we could.
@hirtie-maxim Are these @houndci complaints fixable? I currently cannot check as I am sitting in the ubahn :laughing:

**Comment by 0xF013** [comment]
@floriank I was making a joke on my test-handling technique.

**Comment by floriank** [comment]
I need an `irony` smiley.

Once again: summoning the presence of @hirtie-maxim - what about hound? I reviewed that PR and it looks good so far - I'd only like the code style updated accordingly.

**Comment by knight-bubble** [comment]
@floriank, should it be updated for specs? Tried to add ignore comments, or to add the specs to jshintignore file, but without effect.

**Comment by floriank** [comment]
Yes, specs and tests should follow the conventions.

**Comment by 0xF013** [comment]
@floriank some of protractor instructions are a funny DSL that makes hound go rabid.

**Comment by floriank** [comment]
@0xF013 didin't know that! I suppose it's okay then - thanks for the time investment _makesMentalNote_

**Comment by floriank** [comment]
the `z-index` on the datepicker widget is currently broken for me in chrome:

screenshot from 2015-03-25 16 32 09

upping it shows the widget again:

screenshot from 2015-03-25 16 32 50

Also, I believe the "no <start/end> date set" should be capitalized.

**Comment by ulferts** [comment]
The cuke failure is a query_menu_items failure and as such, it can be ignored.

**Comment by ulferts** [comment]
I am ignoring the styling flaws as those will be addressed separately. This leaves the following I noticed:

## No ... date

There is "no ... date" displayed in the edit fields when there shouldn't and the value of the date picker is then set to the current date:

image

clicking on the field:
image

And the date picker on the due date:
image

## Placement when opening to the top

When not enough space is available below, the date picker opens to the top (good). But the placement is off then:
image

## "No ... date" not placeholder attribute

The "No ... date" should only be a placeholder so that when I start typing I don't have to remove the whole text (is done automatically when using the picker.

## Space between start and end date:

Currently, there is no space between the dash and the start and end date. This makes it hard to read especially if the format is DD-MM-YYYY
image

## Error when having format set to anything but DD/MM/YYYY

The format chosen via the administrative settings (Admin -> Settings -> Display) is correctly displayed in read mode. But when trying to update the field then, errors are returned as the request to the server is no longer constructed correctly:
image
Console for that:
image

Or:
image
Console for that:
image

**Comment by knight-bubble** [comment]
@ulferts, You commented:
I checked with PM and the requirement is to always display the dates in the format YYYY/MM/DD in while in edit mode. When in show mode, the display should be according to the settings in the admin section. Currently this admin setting is used for both which I personally find more desirable.

So should be the date format chosen in edit mode?

**Comment by floriank** [comment]
@hirtie-maxim I believe @ulferts talked about personal preference here. I concur with that opinion - I would say that the display in edit mode, as well as in read mode should follow the configured format in the settings. 

I would also consider this a detail - if it takes too much time to change it so that the edit mode always uses the ISO8601 Format `YYYY-MM-DD` and the read mode uses the configured version, then please do not spend time on this detail.

By the way, @ulferts is also out of commission for 2 months and will probably not respond.

**Comment by floriank** [comment]
@hirtie-maxim Please remove the WIP label once you consider this mergable.

**Comment by knight-bubble** [comment]
This pull request will be cherry-picked by @0xF013, once he is done with his huge PR.

**Comment by floriank** [comment]
Ah, so it was this one - we were unsure a moment ago.

**Comment by NobodysNightmare** [comment]
@hirtie-maxim Please rebase your work onto the dev branch.

**Comment by NobodysNightmare** [comment]
Style issues:
date-alignment

The vertical alignment of the label (**Date**) and the date range picker is off.
- **red**: The baselines do not match (_good_ baselines of other elements also highlighted)
- **pink**: The label is not centered w.r.t. the input fields

**Comment by NobodysNightmare** [comment]
There is no edit control for custom fields of type `Date`.

Is this intended?

**Comment by NobodysNightmare** [comment]
When scrolling the date picker stays at its absolute position and does not scroll with the pane, while other elements (like the edit controls) scroll with the pane:

date-picker-scroll

**Comment by NobodysNightmare** [comment]
You are now hiding the date picker on scrolling...
This means that after scrolling I have to deselect the input field and select it again to make the date picker visible again.

Isn't it possible to keep it open, but aligned with the scrolling of the details pane (like the edit controls do)?

**Comment by NobodysNightmare** [comment]
I think the date picker should be hidden after clicking `Clear`, as it is otherwise hard to submit the changes, because the picker overlaps the edit controls.

Hiding the picker after clicking "clear" would also be consistent with the fact that the picker is hidden after you click on a date.

**Comment by NobodysNightmare** [comment]
The travis build is failing because you commited chagnes to the `Gemfile.lock`

**Comment by NobodysNightmare** [comment]
We want to merge this earlier, though not "perfect".

I will re-issue that PR with a reverted `Gemfile.lock` and ideally get that PR merged than.

For the remaining issues that arose during this PR I will create separate work packages.

**Comment by NobodysNightmare** [comment]
closing in favor of https://github.com/opf/openproject/pull/2816

**Pull request closed without merging**